Applications/Spire/Source/OrderImbalanceIndicator/CurrentOrderImbalanceIndicatorTableModel.cpp

```diff
@@ -0,0 +1,139 @@
+#include "Spire/OrderImbalanceIndicator/CurrentOrderImbalanceIndicatorTableModel.hpp"
+#include <Beam/Queues/Publisher.hpp>
+#include <Beam/Threading/Timer.hpp>
+
+using namespace Beam;
+using namespace Beam::Threading;
+using namespace Beam::TimeService;
+using namespace boost;
+using namespace boost::gregorian;
+using namespace boost::posix_time;
+using namespace boost::signals2;
+using namespace Nexus;
+using namespace Spire;
+
+CurrentOrderImbalanceIndicatorTableModel::CallbackTimer::CallbackTimer(
+    std::unique_ptr<TimerBox> timer, std::function<void ()> expiration_callback)
+    : m_timer(std::move(timer)),
+      m_expiration_callback(std::move(expiration_callback)) {
+  m_timer->GetPublisher().Monitor(
+    m_queue.GetSlot<Timer::Result>([=] (auto result) {
+      if(result == Timer::Result::EXPIRED) { m_expiration_callback(); }
+    }));
+  m_timer->Start();
+}
+
+CurrentOrderImbalanceIndicatorTableModel::CurrentOrderImbalanceIndicatorTableModel(
+    time_duration offset, TimeClientBox clock, TimerFactory timer_factory,
+    std::shared_ptr<OrderImbalanceIndicatorModel> source)
+    : m_source(std::move(source)),
+      m_offset(offset),
+      m_clock(std::move(clock)),
+      m_timer_factory(std::move(timer_factory)) {
+  m_operation_connection = m_table.connect_operation_signal(std::bind_front(
+    &CurrentOrderImbalanceIndicatorTableModel::on_operation, this));
+  auto subscription = m_source->subscribe(
+    TimeInterval::closed(m_clock.GetTime() - m_offset,
+      std::numeric_limits<ptime>::max()), std::bind_front(
+    &CurrentOrderImbalanceIndicatorTableModel::on_imbalance, this));
+  m_subscription_connection = std::move(subscription.m_connection);
+  m_load = subscription.m_snapshot.then(
+    [=] (auto&& result) { on_load(result.Get()); });
```

Comment:
Why is this being passed as a forwarding reference?

Reply:
I'd like to avoid copying the result (which could be large), but the value returned by the promise will be a temporary so I can't take it by reference. My understanding of this is that result's lifetime is equal to it's use, so it will persist until the call to on_load has completed.

Reply:
Follow up on the rules for how references work and then fix this up. As is it's too inconsistent and I'm not sure what the purpose of it is.